[cpu] implement scaled dot product flash attention

Valentine233 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105131
* #107128
* #104863
* #104693
* __->__ #103826
* #104584
* #104583

Feature RFC: https://github.com/pytorch/rfcs/pull/56.

The flash attention CPU kernel is added, for forward path FP32. Blocking is applied on dimensions of query length and kv length and the fusion of gemm + softmax update + gemm is done at once for each block. Parallelization is on the dimensions of batch size, head number and query length. In addition, the causal attention mask is supported. As the attention is masked for the unseen tokens, early termination is applied and we only calculate the blocks in the lower triangular part.

cc @jgong5 @mingfeima @EikanWang @XiaobingSuper @sanchitintel @ashokei @jingxu10

Valentine233 committed:
[cpu] implement scaled dot product flash attention

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,512 @@
+#include <ATen/ops/empty.h>
+#endif
+
+#if AT_MKL_ENABLED()
```
later on we can remove `mkl.h` from this file, now you can leave it as it is.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,512 @@
+
+namespace {
+
+using fVec = vec::Vectorized<float>;
```
place `using fVec = vec::Vectorized<float>;` in function, do not use global static variable.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+using fVec = vec::Vectorized<float>;
+static auto f_vec_size = fVec::size();
+
+inline void _store(
```
we can use `map` from functional to write most of the logic, it would be more clean. as long as the buffer size is L1 hit, the perf doesn't make much difference.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+    fVec src,
+    int64_t l=fVec::size()) {
+  auto res = vec::convert_float_bfloat16(src, src);
+  res.store(dst, l);
```
store when size != vec::size() is slow. it has an addional memcpy.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+  res.store(dst, l);
+}
+
+inline void _mkl_gemm(
```
this shall be gone in future.

cpublas::gemm has some additional checks and normalized the 0 stride (some corner cases when tensor is broadcasted). directly go into mkl would have segfault.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+}
+
+template <typename scalar_t>
+inline void _normalization_kernel(
```
I will write an overload when the out has different dtype with input in `map`

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+      out, a, size);
+}
+
+const std::vector<int64_t> qsplit_range{767, 191, 31};
```
do not use global variables.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+const int64_t kvsplit_size = 512;
+
+template <typename scalar_t>
+at::ScalarType ConvertTypeToKType() {
```
this did not cover other dtypes such as double.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+    const int& kvsize,
+    const int& headsize,
+    const int& idx) {
+  float tmp_max = 0.f, tmp_sum = 0.f, sum_old = 0.f, exp_tmp = 0.f;
```
try to `using accum_t  = at::opmath_type<scalar_t>;` here.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+
+    for (int64_t x = begin; x < end; x++) {
+      int qBlockSize = (k == qSlice - 1) ? qTail : qSplitSize;
+      int ompIdx = at::get_thread_num();
```
move this next to `data_index_init`, this is a slow function call.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+      int qBlockSize = (k == qSlice - 1) ? qTail : qSplitSize;
+      int ompIdx = at::get_thread_num();
+      _init_mha_buffer_kernel(
+          qk_max.data_ptr<float>() + ompIdx * qSplitSize,
```
`data_ptr` is a un-inlined function call, should be placed at inner loop since it is slow.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+  // Query -> Query(Batch x Q_seq_len x hiddenSize)
+  // Key   -> Key(Batch x KV_seq_len x hiddenSize)
+  // Value -> Value(Batch x KV_seq_len x hiddenSize)
+  Tensor q = query.transpose(1, 2).reshape(
```
technically it is possible that this `reshape` may fail when `num_head` and `head_size` are not contiguous dimension.
this is not common case but consider remove the reshape.

mingfeima commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,516 @@
+  const auto softmax_scale =
+      sdp::calculate_scale(query, scale).as_float_unchecked();
+
+  if (query.scalar_type() == at::kFloat) {
```
the standard approach here is to use `AT_DISPATCH_FLOATING_TYPES_AND(ScalarType::BFloat16, query.scalar_type(),`

mingfeima reviewed (changes requested):
just run some easy to obvious issues, still this patch may need more attention.

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,410 @@
+  int64_t qSize = query.size(1);
+  int64_t kvSize = value.size(1);
+  int64_t num_head = query.size(2);
+  int64_t headSize = query.size(3);
```
Thanks, updated in PR [add sdpa choice and UT](https://github.com/pytorch/pytorch/pull/105131).

Valentine233 commented on aten/src/ATen/native/cpu/FlashAttentionKernel.cpp:
```diff
@@ -0,0 +1,410 @@
+  int64_t qSlice = (qSize - 1) / qSplitSize + 1;
+  int64_t num_thread = at::get_num_threads();
+
+  at::Tensor qk = at::empty({num_thread, qSplitSize, kvSplitSize}, at::kFloat);
```
Thanks, updated in PR [enable bfloat16 and refactor for flash attention](https://github.com/pytorch/pytorch/pull/104863).

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"




cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

drisspg reviewed (approved):
Going to work my way up the rest of the stack but I think this makes sense, first thought is can we add arbitrary mask support but will continue reading up to see what has changed

Valentine233 commented:
> Going to work my way up the rest of the stack but I think this makes sense, first thought is can we add arbitrary mask support but will continue reading up to see what has changed

Thanks for your reply. For now, the arbitrary mask is not supported in flash attention and this would be added later.

drisspg commented:
@Valentine233  `_scaled_dot_product_efficient_attention` is the aten function that currently accepts an optional attn_bias. Since flash_attention does not support this feature it doesn't have that. If the CPU kernel can support this it might be easier to register this new function under the efficient_attention name.

Valentine233 commented:
> @Valentine233 `_scaled_dot_product_efficient_attention` is the aten function that currently accepts an optional attn_bias. Since flash_attention does not support this feature it doesn't have that. If the CPU kernel can support this it might be easier to register this new function under the efficient_attention name.

Do you mean that for CPU, we do flash attention **WITHOUT** attn_mask in `_scaled_dot_product_flash_attention` and do flash attention **WITH** attn_mask in `_scaled_dot_product_efficient_attention`?

In addition, I am confused why `_scaled_dot_product_flash_attention` does not support this feature.

drisspg commented:
So currently the only aten function that can calculate SDPA on cpu is _scaled_dot_product_math(). So Flash/MemEff are specifically written to cater to what the cuda implementations support. FlashAttentionV1 and V2 do not support passing in an arbitrary attention bias. Where as of recently: https://github.com/pytorch/pytorch/pull/104310, the kernel does support adding an arbitrary attention bias.

drisspg commented:
Looking at the CPU implementation it doesn't look there is support for passing an arbitrary attention bias, which is fine just wasn't sure if there was a plan to support this feature

Valentine233 commented:
Sure! In conclusion, if CPU want to add the arbitrary attention bias for flash attention in the future, it should be add in `_scaled_dot_product_efficient_attention`.

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"


Feature RFC: https://github.com/pytorch/rfcs/pull/56.

The flash attention CPU kernel is added, for forward path FP32. Blocking is applied on dimensions of query length and kv length and the fusion of gemm + softmax update + gemm is done at once for each block. Parallelization is on the dimensions of batch size, head number and query length. In addition, the causal attention mask is supported. As the attention is masked for the unseen tokens, early termination is applied and we only calculate the blocks in the lower triangular part.

cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 committed:
Update on "[cpu] implement scaled dot product flash attention"


Feature RFC: https://github.com/pytorch/rfcs/pull/56.

The flash attention CPU kernel is added, for forward path FP32. Blocking is applied on dimensions of query length and kv length and the fusion of gemm + softmax update + gemm is done at once for each block. Parallelization is on the dimensions of batch size, head number and query length. In addition, the causal attention mask is supported. As the attention is masked for the unseen tokens, early termination is applied and we only calculate the blocks in the lower triangular part.

cc jgong5 mingfeima EikanWang XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

Valentine233 commented:
@pytorchbot merge
